File: python/003_longest_substring_without_repeating_characters.py
def lengthOfLongestSubstring(s: str):
    if len(s) == 0:
        return 0
    max_l = 0
    head = 0
    tail = head + 1
    # cur_l = {}
    while tail < len(s):
        # pos = cur_l.get(s[tail], -1)
        pos = s.find(s[tail], head, tail)
        if pos != -1:
            max_l = max(max_l, tail - head)
            head = pos + 1
            # head = max(head, pos + 1)
        # cur_l[s[tail]] = tail
        tail = tail + 1
    max_l = max(max_l, tail - head)
    return max_l


print(lengthOfLongestSubstring('abba'))

File: cpp/003_longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s == "") {
            return 0;
        }

        if (s.length() == 1) {
            return 1;
        }

        unordered_map<char, int> dic;
        dic[s[0]] = 0;
        int min = 0;
        int max = 0;
        int result = 1;
        for(int i = 1; i<s.length(); i++) {
            auto it = dic.find(s[i]);
            if (it == dic.end()) {
                dic[s[i]] = i;
            } else {
                // found
                if ((max - min + 1) > result) {
                    result = max -min +1;
                }
                if ((dic[(s[i])] + 1) > min)
                    min = dic[(s[i])] + 1;
                dic[(s[i])] = i;
            }
            max = i;
        }
        if ((max - min + 1) > result) {
            result = max -min+1;
        }
        return result;
    }
};
